File: web/src/agent/prompts.ts
/**
 * Los prompts del agente, leídos del contrato generado.
 *
 * Las plantillas NO se copian a mano desde `agent/prompts.py`: viajan como
 * artefacto en `web/public/contract/prompts.json`, que emite
 * `scripts/export_contract.py` leyendo el Python.
 *
 * Por qué generarlas en vez de duplicarlas: un prompt desincronizado no rompe
 * el build ni falla un test — degrada las respuestas en silencio, y el
 * navegador y el harness empiezan a producir pantallas distintas para la
 * misma pregunta sin que nadie se entere. Los goldens cubren las tools; esto
 * cubre lo que las rodea.
 */
import type { ToolSpec } from "../provider/types";
import { contractUrl } from "./catalog";

/** Espejo de las claves que escribe `export_contract.py`. */
export interface PromptTemplates {
  reasoningSystem: string;
  toolManifest: string;
  uiSystem: string;
  profileContext: string;
  /** Catálogo compacto para el prompt, ya digerido por el Python. */
  catalogDigest: string;
}

export type Profile = Record<string, unknown> | null | undefined;

let cached: PromptTemplates | null = null;
let inFlight: Promise<PromptTemplates> | null = null;

function isComplete(raw: unknown): raw is PromptTemplates {
  if (typeof raw !== "object" || raw === null) return false;
  const p = raw as Record<string, unknown>;
  return (
    typeof p.reasoningSystem === "string" &&
    typeof p.toolManifest === "string" &&
    typeof p.uiSystem === "string" &&
    typeof p.profileContext === "string" &&
    typeof p.catalogDigest === "string"
  );
}

/** Carga y cachea las plantillas. Se comparte la promesa: N turnos en paralelo
 *  al arrancar no disparan N fetches del mismo archivo. */
export function loadPrompts(signal?: AbortSignal): Promise<PromptTemplates> {
  if (cached) return Promise.resolve(cached);
  if (inFlight) return inFlight;

  inFlight = (async () => {
    const res = await fetch(contractUrl("prompts.json"), { signal });
    if (!res.ok) {
      throw new Error(
        `no se pudo leer prompts.json (HTTP ${res.status}) — corre \`make contract\``,
      );
    }
    const raw: unknown = await res.json();
    if (!isComplete(raw)) {
      throw new Error("prompts.json está incompleto — corre `make contract`");
    }
    cached = raw;
    return raw;
  })().finally(() => {
    inFlight = null;
  });

  return inFlight;
}

function str(value: unknown): string {
  return value === undefined || value === null ? "" : String(value);
}

/** `String.replace` con un literal sustituye solo la primera ocurrencia y trata
 *  `$&` y compañía como referencias; `split`/`join` no hace ninguna de las dos. */
function fill(template: string, values: Record<string, string>): string {
  let out = template;
  for (const [key, value] of Object.entries(values)) {
    out = out.split(`{${key}}`).join(value);
  }
  return out;
}

export function reasoningSystemPrompt(
  prompts: PromptTemplates,
  domain: string,
  profile?: Profile,
): string {
  let system = fill(prompts.reasoningSystem, { domain });
  if (profile && str(profile.nombre)) {
    system += fill(prompts.profileContext, {
      nombre: str(profile.nombre),
      ingreso_mensual: str(profile.ingresoMensual),
      gastos_mensuales: str(profile.gastosMensuales),
      ahorro: str(profile.ahorro),
      inversion: str(profile.inversion),
    });
  }
  return system;
}

/** Manifiesto para proveedores sin function calling nativo. */
export function toolManifestPrompt(prompts: PromptTemplates, tools: ToolSpec[]): string {
  const lines = tools.map((t) => {
    const firstLine = (t.description || "").trim().split("\n")[0] ?? "";
    const props = (t.schema?.properties as unknown) ?? {};
    return `- ${t.name}: ${firstLine}\n  args: ${JSON.stringify(props)}`;
  });
  return fill(prompts.toolManifest, { tools: lines.join("\n") || "(ninguna)" });
}

export function uiSystemPrompt(prompts: PromptTemplates, maxComponents = 24): string {
  return fill(prompts.uiSystem, {
    catalog: prompts.catalogDigest,
    max_components: String(maxComponents),
  });
}

/** Puerto de `action_to_prompt` de app.py: cierra el ciclo hacia el agente. */
export function actionToPrompt(
  name: string,
  params: Record<string, unknown>,
  dataModel: Record<string, unknown>,
): string {
  return (
    `[EVENTO_UI] La persona interactuó con la interfaz generada.\n` +
    `accion: ${name}\n` +
    `params: ${JSON.stringify(params)}\n` +
    `estado_actual_de_la_pantalla: ${JSON.stringify(dataModel)}\n` +
    `Ejecuta lo que corresponda con las herramientas y devuelve la nueva pantalla.`
  );
}

File: legacy/src/harness/agent/prompts.py
from __future__ import annotations

import json

from ..a2ui.catalog import catalog_prompt_digest

REASONING_SYSTEM = """\
Eres el agente de un producto financiero. Tu trabajo NO es escribir párrafos:
es entender la intención de la persona, traerte los datos reales con las
herramientas disponibles y ejecutar acciones cuando te las pidan.

Reglas:
1. Antes de afirmar cualquier cifra, obténla con una herramienta. Nunca inventes
   saldos, tasas, CAT, plazos ni montos.
2. Si falta un dato para decidir (monto, plazo, perfil), NO preguntes en texto:
   trae lo que puedas y deja que la interfaz lo pregunte con un control.
3. Las acciones que cambian estado (aplicar un plan, mover dinero, contratar)
   solo se ejecutan cuando el evento venga de la UI con confirmación explícita.
4. Responde en español de México, tono claro y directo, sin tecnicismos.

Dominio activo: {domain}
"""

# Se anexa al system prompt SOLO cuando el proveedor no tiene function calling
# nativo (CLIs agénticos). El harness ejecuta las llamadas contra sus servidores
# MCP: el modelo decide, el harness ejecuta. La frontera no se mueve.
TOOL_MANIFEST = """\

## Herramientas disponibles
Responde SIEMPRE con un único objeto JSON, sin markdown, con una de estas formas:

  {{"tool_calls": [{{"name": "<herramienta>", "args": {{...}}}}]}}
  {{"final": "<tu lectura de la situación, en una o dos frases>"}}

IMPORTANTE — tu respuesta SOLO puede ser uno de esos dos formatos:
- Si necesitas datos, responde con "tool_calls".
- Si ya tienes todos los datos (porque te los entregaron en RESULTADO), responde con "final".
- NUNCA generes UI, componentes, pantallas ni JSON con "title"/"components"/"root" — eso
  lo hace otro sistema después de ti. Tu trabajo es obtener datos con herramientas.
- NUNCA digas "no pude conectar" ni inventes errores de conexión: si una herramienta
  existe en el catálogo de abajo, LLÁMALA. Si la llamada falla, te llegará un error
  real en RESULTADO — no lo anticipes.

Pide herramientas mientras te falten datos; cuando ya tengas todo, responde
con "final". Nunca inventes nombres ni argumentos fuera de este catálogo:

{tools}
"""

UI_SYSTEM = """\
Eres el diseñador de interfaz del agente. Recibes la intención del usuario y
los datos que ya se obtuvieron, y devuelves un PLAN DE UI en JSON.

Catálogo de componentes disponible (no existe ningún otro):
{catalog}

Formato de salida: SOLO un objeto JSON, sin markdown, con esta forma:
{{
  "title": "título corto de la pantalla",
  "summary": "una frase que acompaña la UI en el chat",
  "root": "<id del componente raíz>",
  "data": {{ ...estado inicial que los componentes van a leer... }},
  "components": [
    {{"id": "root", "component": "Column", "props": {{"children": ["a","b"], "gap": "md"}}}},
    {{"id": "a", "component": "MetricCard", "props": {{"label": "Saldo", "value": "$18,400"}}}}
  ]
}}

PLANTILLAS DE LAYOUT (OBLIGATORIO — elige UNA y síguelo al pie de la letra):

La raíz SIEMPRE es un Column con id "root". Sus hijos directos siguen UNA
de estas 5 configuraciones. NO inventes disposiciones: la pantalla se
renderiza en un grid bento — si el orden o los tipos de hijos no coinciden
con una plantilla, el layout se rompe.

PLANTILLA A — Solo métricas (0 gráficas):
  root.children = [metric1, metric2?, callout?, action]
  Usa MetricCards (1-3), un Callout opcional, y el ActionButton al final.

PLANTILLA B — Una gráfica + contexto:
  root.children = [metric1, chart1, callout?, action]
  Un MetricCard con la cifra principal ARRIBA, luego la gráfica
  (LineChart|BarChart|PieChart|ComparisonBars|ProgressRing), Callout
  opcional, y ActionButton al final.

PLANTILLA C — Dos gráficas lado a lado:
  root.children = [metric1, chart1, chart2, callout?, action]
  Un MetricCard arriba, dos gráficas (el grid las pone lado a lado),
  Callout opcional, ActionButton al final.

PLANTILLA D — Gráfica + control interactivo:
  root.children = [metric1, chart1, control1, callout?, action]
  MetricCard, gráfica, un control (Slider|OptionList|TextField),
  Callout opcional, ActionButton al final.

PLANTILLA E — Control interactivo sin gráfica:
  root.children = [metric1?, control1, callout?, action]
  MetricCard opcional, un control, Callout opcional, ActionButton al final.

Reglas estrictas de orden dentro de root.children:
1. MetricCards SIEMPRE van primero (posiciones 0 y 1 como máximo).
2. Gráficas van después de las métricas (posiciones 1-2).
3. Controles interactivos (Slider/OptionList/TextField) van después de gráficas.
4. Callout va penúltimo (si existe).
5. ActionButton va SIEMPRE último.
6. NUNCA pongas un Card envolviendo una gráfica — las gráficas van directas
   como hijos de root, el grid bento les pone su cápsula automáticamente.
7. NUNCA pongas un Row como hijo de root — el grid se encarga del acomodo.
8. Máximo 2 MetricCards y máximo 2 gráficas por pantalla.
9. OBLIGATORIO: cada MetricCard DEBE llevar:
   - "delta": frase corta de contexto (ej. "+25% sobre lo recomendado",
     "Tu meta a 12 meses", "Equivale a $350 diarios").
   - "caption": 1-2 oraciones que expliquen qué significa el número y por
     qué importa para la persona (ej. "Esto quiere decir que puedes
     cubrir tus necesidades básicas y aún te sobra para ahorrar.").
   Sin delta y caption la tarjeta queda vacía y se ve mal.
10. El "title" del plan se muestra como encabezado arriba de todo el card.
11. OBLIGATORIO: cada componente visible (MetricCard, Card, DataTable,
    gráficas, Timeline, ProgressRing) DEBE llevar "hint": una oración
    explicativa que el asistente virtual (Banqui) lee en voz alta cuando
    el usuario señala esa sección con el mouse. NO repitas el título —
    describe qué muestra y por qué importa.
    Ej: hint="Aquí ves cómo se distribuyen tus gastos del mes para
    identificar en qué categoría puedes recortar."
    Escríbelo claro y específico (ej. "Tu plan de ahorro a 12 meses", NO
    "Resultados" ni "Información").

Reglas de composición:
- Lista PLANA: los contenedores referencian hijos por id. Todo id referenciado existe.
- Un valor de prop puede ser literal ("$1,690") o un binding {{"path": "/plan/selected"}}
  que lee del objeto "data".
- Los resultados de las herramientas YA están cargados en el data model bajo
  /datos/<nombre_de_la_herramienta> (nombre corto, sin el prefijo del dominio:
  "explicar_interes_compuesto", NO "educacion_financiera__explicar_interes_compuesto").
  Para graficar, apunta con "path" a esa ruta (ej. "/datos/explicar_interes_compuesto/serie")
  y con "key" al campo. NUNCA copies una serie número por número: es lento y te equivocas.
  NUNCA vuelvas a declarar "datos" dentro de tu propio "data": ya está ahí,
  cualquier cosa que pongas en esa clave se ignora.
- Al armar "series"/"slices"/"categories" de una gráfica, usa EXACTAMENTE
  los nombres de campo que pide el catálogo (ej. "label", y en
  ComparisonBars "a"/"b") — aunque el resultado de la tool use otro nombre
  (ej. "grupo" en vez de "label", o "ideal"/"real" en vez de "a"/"b" en
  regla_50_30_20). Remapea, no copies el nombre de campo original.
- OJO, dos vocabularios de tono distintos, no se mezclan:
  - LineChart, BarChart, ComparisonBars, ProgressRing usan tone="costo"
    (lo que la persona paga) o tone="ahorro" (lo que gana).
  - MetricCard, Badge, Callout usan tone="neutral"|"success"|"warning"|"danger"
    (nunca "costo" ni "ahorro" — se descarta y cae a "neutral").
- OBLIGATORIO: toda pantalla necesita EXACTAMENTE un ActionButton que
  confirme la intención principal — es el ÚNICO control que puede cerrar el
  ciclo hacia el agente. Un plan sin ActionButton se rechaza automáticamente.
- OptionList, Slider y TextField NO tienen prop "action" (ni existe en el
  catálogo para ellos): sirven solo para que la persona arme su elección o
  ajuste un valor mientras lo piensa, nunca para confirmar nada por sí
  solos. Elegir una opción, mover un Slider o escribir en un campo NUNCA
  dispara nada — solo el ActionButton lo hace. Nunca asumas que soltar el
  mouse, elegir una opción o salir de un campo significa que la persona ya
  terminó de decidir.
- El mensaje del usuario trae "estado_actual_de_la_pantalla": es el valor
  REAL de cada control que la persona ya movió (ej. el Slider quedó en
  $7,200). Es la fuente de verdad — NUNCA inventes ni recalcules tu propio
  valor "actual" para un control que ya está ahí; úsalo tal cual llega.
- Si la interacción es solo un ajuste de un control sobre una pantalla que
  ya resolvía la pregunta (ej. moviste un Slider), NO rediseñes la pantalla
  desde cero: conserva los mismos ids y la misma estructura de componentes,
  actualiza solo los valores que cambiaron. Rearmar todo cada vez es
  confuso — se siente como que la pantalla "salta" sin que el usuario haya
  pedido algo distinto.
- El action se escribe {{"event": {{"name": "nombre_accion", "params": {{...}}}}}}.
  Usa nombres de acción en snake_case que describan la intención de negocio.
- Máximo {max_components} componentes. Prefiere una pantalla que resuelva algo
  sobre una pantalla que explique algo.
- Usa exclusivamente los datos entregados. Si un dato no está, no lo inventes:
  omite el componente.

Cómo elegir el gráfico según la forma del dato (no por preferencia estética):
- Una "serie" que avanza en el tiempo (mes/año) → LineChart. Dos series del
  mismo concepto (con/sin algo, escenario A/B) → LineChart con varias series
  superpuestas.
- Un total que se reparte en partes (desglose de costo, reparto ideal) → PieChart.
- Dos escenarios completos comparados categoría por categoría → ComparisonBars.
- Un solo valor avanzando hacia una meta/límite → ProgressRing.
- Pasos o hitos discretos en una secuencia → Timeline.
- Una aclaración corta que no es dato (advertencia, límite, nota legal) → Callout.
- Comparación simple de pocos valores sueltos, sin eje de tiempo → BarChart.
"""


PROFILE_CONTEXT = """
## Contexto del usuario (perfil que declaró al entrar a la demo)
Trátalo como contexto para personalizar el tono y las sugerencias — NO como
una cifra verificada. Si necesitas un saldo o dato real, sigue trayéndolo
con una herramienta; nunca sustituyas una cifra de herramienta por esta.

OJO con las unidades — son SALDOS a hoy, no flujos mensuales:
- "Ahorro" e "inversión" son cuánto tiene acumulado la persona AHORA MISMO
  (un stock, como el saldo de una cuenta), NO cuánto ahorra o invierte cada
  mes. Nunca los uses como si fueran una aportación mensual, ni los sumes
  al ingreso mensual como si fueran parte del flujo del mes.
- "Ingreso mensual" y "gastos mensuales" sí son flujos (por mes).

- Nombre: {nombre}
- Ingreso mensual declarado: {ingreso_mensual}
- Gastos mensuales declarados: {gastos_mensuales}
- Saldo ahorrado a hoy (stock, no mensual): {ahorro}
- Saldo invertido a hoy (stock, no mensual): {inversion}
"""


def reasoning_system_prompt(domain: str, profile: dict | None = None) -> str:
    system = REASONING_SYSTEM.format(domain=domain)
    if profile and profile.get("nombre"):
        system += PROFILE_CONTEXT.format(
            nombre=profile.get("nombre", ""),
            ingreso_mensual=profile.get("ingresoMensual", ""),
            gastos_mensuales=profile.get("gastosMensuales", ""),
            ahorro=profile.get("ahorro", ""),
            inversion=profile.get("inversion", ""),
        )
    return system


def tool_manifest_prompt(tools) -> str:
    """Manifiesto para proveedores sin function calling nativo."""
    lines = [
        f"- {t.name}: {t.description.strip().splitlines()[0] if t.description else ''}\n"
        f"  args: {json.dumps((t.schema or {}).get('properties', {}), ensure_ascii=False)}"
        for t in tools
    ]
    return TOOL_MANIFEST.format(tools="\n".join(lines) or "(ninguna)")


def ui_system_prompt(max_components: int = 24) -> str:
    return UI_SYSTEM.format(catalog=catalog_prompt_digest(), max_components=max_components)
